pkg/audit/manager.go

```diff
@@ -274,62 +275,107 @@ func (am *Manager) auditResources(
 	var errs opa.Errors
 	nsCache := newNSCache()
 
-	for gv, gvKinds := range clusterAPIResources {
-		for kind := range gvKinds {
-			objList := &unstructured.UnstructuredList{}
-			opts := &client.ListOptions{
-				Limit: int64(*auditChunkSize),
+	constraintsList := make(map[string]bool)
+	if *auditMatchKindOnly {
+		constraintList := &unstructured.UnstructuredList{}
+		for _, c := range resourceList {
+			constraintList.SetGroupVersionKind(c)
+			if err = am.client.List(ctx, constraintList); err != nil {
+				am.log.Error(err, "Unable to list objects for gvk", "group", c.Group, "version", c.Version, "kind", c.Kind)
+				continue
 			}
-			resourceVersion := ""
-
-			for {
-				objList.SetGroupVersionKind(schema.GroupVersionKind{
-					Group:   gv.Group,
-					Version: gv.Version,
-					Kind:    kind + "List",
-				})
-				objList.SetResourceVersion(resourceVersion)
-
-				err := am.client.List(ctx, objList, opts)
+			for _, constraint := range constraintList.Items {
+				kinds, found, err := unstructured.NestedSlice(constraint.Object, "spec", "match", "kinds")
 				if err != nil {
-					am.log.Error(err, "Unable to list objects for gvk", "group", gv.Group, "version", gv.Version, "kind", kind)
+					am.log.Error(err, "Unable to return spec.match.kinds field", "group", c.Group, "version", c.Version, "kind", c.Kind)
 					continue
 				}
+				if found {
+					for _, k := range kinds {
+						kind, ok := k.(map[string]interface{})
+						if !ok {
+							am.log.Error(errors.New("could not cast kind as map[string]"), "kind", k)
+							continue
+						}
+						kindsKind, _, err := unstructured.NestedSlice(kind, "kinds")
+						if err != nil {
+							am.log.Error(err, "Unable to return kinds.kinds field", "group", c.Group, "version", c.Version, "kind", c.Kind)
+							continue
+						}
+						for _, kk := range kindsKind {
+							// adding constraint match kind to list
+							constraintsList[kk.(string)] = true
```

Comment:
we should probably lookup by a combination of gvk, e.g. `g-v-k`

Reply:
we are looping over `kinds` here: https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/audit/manager.go#L278
Wouldn't that check against all groups of that kind?

Reply:
We are iterating over all groups and then over all kinds in that group.

We should probably be skipping all iterations where either group or kind doesn't match in order to maximize the number of resources excluded. It does make how to deal with starred groups harder though. We'll probably need nested maps.

Reply:
Is this worth the complexity? Is there any case other than Ingress?

Reply:
> check against all groups of that kind

I'm ok with the current implementation to reduce complexity.

Reply:
The upper bound is theoretically infinity, as users can create infinite "Ingress" CRDs belonging to different groups. As a practical matter, there will still be a benefit.

The risk is if there becomes some super-generic popular Kind that multiple groups use.